micrometer-shim/src/main/java/io/opentelemetry/contrib/metrics/micrometer/DoubleCounter.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package io.opentelemetry.contrib.metrics.micrometer;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.Tags;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.metrics.BoundDoubleCounter;
+import io.opentelemetry.api.metrics.DoubleCounterBuilder;
+import io.opentelemetry.api.metrics.LongCounterBuilder;
+import io.opentelemetry.api.metrics.ObservableDoubleMeasurement;
+import io.opentelemetry.context.Context;
+import java.util.function.Consumer;
+
+final class DoubleCounter implements io.opentelemetry.api.metrics.DoubleCounter {
+  private final SharedMeterState state;
+  private final Reader<Counter.Builder> factory;
+
+  private DoubleCounter(SharedMeterState state, Reader<Counter.Builder> factory) {
+    this.state = state;
+    this.factory = factory;
+  }
+
+  @Override
+  public void add(double value) {
+    add(value, Attributes.empty(), Context.current());
+  }
+
+  @Override
+  public void add(double value, Attributes attributes) {
+    add(value, attributes, Context.current());
+  }
+
+  @Override
+  public void add(double value, Attributes attributes, Context context) {
+    BoundDoubleCounter bound = bind(attributes);
+    try {
+      bound.add(value, context);
+    } finally {
+      bound.unbind();
+    }
+  }
+
+  @Override
+  public BoundDoubleCounter bind(Attributes attributes) {
+    Tags tags = TagUtils.attributesToTags(attributes);
+    Counter counter = factory.get().tags(tags).register(state.meterRegistry());
```

Comment:
How expensive is it for OTEL's API to constantly register counters in the hot path?

Reply:
https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/common/Attributes.java#L25-L27

the Attributes better have a well-behaved equals/hashcode or things will be very bad.

Reply:
`ArrayBackedAttributes` looks like it does, but `AttibutesMap` does not.  It seems the latter is only used by `SpanBuilder` so perhaps it's safe to assume that any attributes passed to these APIs can be used as a key.

Reply:
`AttributesMap` is a `HashMap` so it should be fine.

Reply:
but yes, these APIs won't get getting the attributes from the internals of the span, I don't think.